src/dspbase.jl

```diff
@@ -186,17 +186,45 @@ conv2(A::StridedMatrix{T}, B::StridedMatrix{T}) where {T<:Integer} =
 conv2(u::StridedVector{T}, v::StridedVector{T}, A::StridedMatrix{T}) where {T<:Integer} =
     round.(Int, conv2(float(u), float(v), float(A)))
 
+function check_mode_kwarg(mode::Symbol, su::Integer, sv::Integer)
+    if mode == :default_full
+        if su != sv
+            Base.depwarn(
+                "The default behavior for xcorr has changed. For more " *
+                "details, see the documentation for xcorr. To avoid this " *
+                "warning, specify mode = :full or mode = :legacy where " *
+                "appropriate",
+                :xcorr
+            )
+        end
+        :full
+    else
+        mode
+    end
+end
+
 """
-    xcorr(u,v)
+    xcorr(u,v; mode = :full)
 
-Compute the cross-correlation of two vectors.
+Compute the cross-correlation of two vectors. The size of the output depends on
+`mode` keyword argument: if `mode = :full` then the result will be the same size
+as [`conv`](@ref) of `u` and `v`. If `mode = :legacy` then the result will have
+length `2*max(length(X), length(Y))-1`, where the beginning of the result will
+be padded with zeros.
 """
-function xcorr(u, v)
+function xcorr(u, v; mode::Symbol = :default_full)
     su = size(u,1); sv = size(v,1)
-    if su < sv
-        u = [u;zeros(eltype(u),sv-su)]
-    elseif sv < su
-        v = [v;zeros(eltype(v),su-sv)]
+    mode = check_mode_kwarg(mode, su, sv)
+    if mode == :legacy
```

Comment:
I agree that this mode shouldn't be the default, but I can see it being a useful feature (as you mentioned before, some folks might want the zero-lag condition to be centered). Maybe `padmode=:longest` or something, rather than calling it `:legacy`.

Reply:
fixed